### PR TITLE
new Activity ValuedParam

### LINKED ISSUE #84: [TASK] Activity-based economic values
--------
Issue Description
--------

Currently to use the activity of a component (e.g. MW produced) for a CashFlow node such as `<driver>`, it is necessary to implement a `Function` `DataGenerator` and extract this data.

However, this would be very straightforward to automate, and simplify users' experiences.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #84 

##### What are the significant changes in functionality due to this change request?
Adds the `<activity>` node, which does a simple lookup of dispatch activity for the associated component.

For example, if a `<Produces>` component consumes `steam` to make `electricity`, a `CashFlow` might include the driver:
```xml
<driver>
  <activity>steam</activity>
  <multiplier>-1</multiplier>
</driver>
```
to suggest a cashflow of -1 * amount of steam consumed (negative * negative makes positive).

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [ ] 4. Automated Tests should pass.
- [ ] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [ ] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
Before this is merged, there's one thing I need to fix, and that's error checking on the resource named.

File: src/ValuedParams/Activity.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      @ In, mode, type of simulation calculation
+      @ In, alias_dict, dict, optional, aliases to use for variable naming
+      @ Out, needs, list, signals needed to evaluate this ValuedParam at runtime
+    """
+    super().read(comp_name, spec, mode, alias_dict=None)
+    # aliases get used to convert variable names, notably for the cashflow's "capacity"
+    if alias_dict is None:
+      alias_dict = {}
+    self._resource = spec.value
+    # FIXME how to confirm this component actually produces/consumes/stores this resource??
```
Comment (PaulTalbot-INL) [inline src/ValuedParams/Activity.py:56]: I'm looking for ideas here; we don't have a reference to the Component, but we need to check  that the named resource is actually produced or consumed by Component. Ideas?

Comment (dylanjm) [reply src/ValuedParams/Activity.py:56]: Are you married to idea of doing the conditional check in this method on read-in? Could we "finalize" a component with conditional check? Or can we interrogate the `spec` variable? Or does `spec` only contain the current xml node information?

Comment (dylanjm) [reply src/ValuedParams/Activity.py:56]: Also to make sure I understand the current dilema, we need to check for user-error cases like this:

```
 <Component name="electr_market">
      <demands resource="steam" dispatch="dependent">
        <capacity>
          <fixed_value>-2</fixed_value>
        </capacity>
      </demands>
      <economics>
        <lifetime>30</lifetime>
        <CashFlow name="e_sales" type="repeating" taxable='True' inflation='none' mult_target='False'>
          <driver>
            <activity>electricity</activity>
            <multiplier>-1</multiplier>
          </driver>
          <reference_price>
      </economics>
    </Component>
```

Where `resource='steam'` conflicts with `<activity>electricity</activity>`

**Review by dylanjm — CHANGES_REQUESTED** [review]
Looks good. Just need to solve the error handling issue.

File: src/ValuedParams/Activity.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply src/ValuedParams/Activity.py:56]: That's true, we could do a "finalize" checking later, as long as it's still in the HERON "compile time" and before RAVEN starts running and producing awful errors. That's a good idea.

Your example is good, yes, that would be an input that I'd expect to see an intelligent error for.

**Review by dylanjm — APPROVED** [review]
All the logic, documentation, and naming checks out. Looks good.

**Pull request merged**